lib/unifex/code_generator/base_types/list.ex

```diff
@@ -27,66 +57,120 @@ defmodule Unifex.CodeGenerator.BaseTypes.List do
           enif_get_list_cell(env, list, &elem, &list);
-          #{
-        BaseType.generate_arg_parse(
-          ctx.subtype,
-          elem_name,
-          ~g<elem>,
-          ctx.postproc_fun,
-          ctx.generator
-        )
-      }
+          #{BaseType.generate_arg_parse(subtype, elem_name, ~g<elem>, postproc_fun, generator)}
         }
       }
       get_list_length_result;
       })
       """
     end
+  end
+
+  defmodule CNode do
+    @moduledoc false
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
 
     @impl BaseType
-    def generate_initialization(name, _ctx) do
-      ~g<#{name} = NULL;>
+    def generate_arg_serialize(name, ctx) do
+      ~g"""
+      ({
+        ei_x_encode_list_header(out_buff, #{name}_length);
+        for(unsigned int i = 0; i < #{name}_length; i++) {
+          #{BaseType.generate_arg_serialize(ctx.subtype, :"#{name}[i]", ctx.generator)}
+        }
+        ei_x_encode_empty_list(out_buff);
+      });
+      """
     end
 
     @impl BaseType
-    def generate_destruction(name, ctx) do
+    def generate_arg_parse(arg, var_name, ctx) do
+      elem_name = :"#{var_name}[i]"
+      len_var_name = "#{var_name}_length"
+      native_type = BaseType.generate_native_type(ctx.subtype, ctx.generator)
+      subtype = ctx.subtype
+      postproc_fun = ctx.postproc_fun
+      generator = ctx.generator
+
       ~g"""
-      if(#{name} != NULL) {
-        for(unsigned int i = 0; i < #{name}_length; i++) {
-          #{BaseType.generate_destruction(ctx.subtype, :"#{name}[i]", ctx.generator)}
+      ({
+        int res = 1;
+        int type;
+        int size;
+        ei_get_type(#{arg}->buff, #{arg}->index, &type, &size);
+        #{len_var_name} = (unsigned int) size;
+        if(type == ERL_LIST_EXT) {
+          res = ei_decode_list_header(#{arg}->buff, #{arg}->index, &size);
+          #{len_var_name} = (unsigned int) size;
+          #{var_name} = malloc(sizeof(#{native_type}) * #{len_var_name});
+
+          for(unsigned int i = 0; i < #{len_var_name}; i++) {
+            #{BaseType.generate_initialization(subtype, elem_name, generator)}
+          }
+
+          for(unsigned int i = 0; i < #{len_var_name}; i++) {
+            #{BaseType.generate_arg_parse(subtype, elem_name, arg, postproc_fun, generator)}
+          }
+        } else if(type == ERL_STRING_EXT) {
+          char *p = malloc(sizeof(char) * #{len_var_name});
+          res = ei_decode_string(#{arg}->buff, #{arg}->index, p);
+          ei_x_buff buff;
+          ei_x_new_with_version(&buff);
+          ei_x_encode_list_header(&buff, #{len_var_name});
+
+          for(unsigned int i = 0; i < #{len_var_name}; i++) {
+            int char_value;
+            if (CHAR_MIN < 0) {
+              char_value = (int)p[i];
+              if(char_value < 0) {
+                char_value = char_value + UCHAR_MAX + 1;
+              }
+            } else {
+              char_value = (int)p[i];
+            }
```

Comment:
I think this can be avoided if you cast `p` to `uint8 *` or `unsigned char *`

Reply:
And I was so proud of this piece of code :cry:  Fixed :+1: